File: LinkedListCycleII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head || !head->next ||!head->next->next)
            return NULL;
        ListNode *slow = head;
        ListNode *fast = head->next->next;
        ListNode *prev = head->next;
        while(fast && slow != fast) {
            if (!fast->next)
                break;
            slow = slow->next;
            prev = fast->next;
            fast = fast->next->next;
        }
        
        if (!fast || !fast->next)
            return NULL;
            
        // fast == slow
        // break the list
        prev->next = NULL;
        
        // calculate each length
        int len1 = len(slow);
        int len2 = len(head);
        advance(slow, len1-len2);
        advance(head, len2-len1);
        
        while (slow != head) {
            slow = slow->next;
            head = head->next;
        }
        
        // restore the list
        prev->next = fast;
        return slow;
        
    }
private:
    int len(ListNode * node) {
        int length = 0;
        while (node) {
            ++length;
            node = node->next;
        }
        return length;
    }
    
    void advance(ListNode * &node, int len) {
        if (len <= 0) return;
        while (len--) {
            node = node->next;
        }
    }
};

File: jiuzhang/linked-list-cycle-ii/linked-list-cycle-ii.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

from lintcode import ListNode

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: the node where the cycle begins. 
                If there is no cycle, return null
    """
    def detectCycle(self, head):
        # write your code here
        if head == None or head.next == None:
            return None
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast == slow:
                break
        if slow == fast:
            slow = head
            while slow != fast:
                slow = slow.next
                fast = fast.next
            return slow
        return None


